File: scripts/github-actions/check_openai_usage.py
#!/usr/bin/env python3
"""OpenAI API使用量チェックスクリプト

トークン使用量を取得し、モデルごとの料金テーブルからコストを試算する。
"""

import os
import sys
from datetime import datetime, timedelta
import requests

OPENAI_ADMIN_KEY = os.environ.get("OPENAI_ADMIN_KEY")
OPENAI_PROJECT_ID = os.environ.get("OPENAI_PROJECT_ID")
SLACK_WEBHOOK_URL = os.environ.get("OPENAI_SLACK_WEBHOOK_URL")
MONTHLY_BUDGET_USD = float(os.environ.get("MONTHLY_BUDGET_USD", "50"))

# 為替レート（円/ドル）
USD_TO_JPY = 150

# gpt-4o-mini料金で統一計算 (USD per 1M tokens)
# https://openai.com/api/pricing/
GPT4O_MINI_PRICING = {"input": 0.15, "output": 0.60}


def get_usage_data(start_timestamp: int, end_timestamp: int) -> dict:
    """トークン使用量を取得"""
    url = "https://api.openai.com/v1/organization/usage/completions"
    params = {
        "start_time": str(start_timestamp),
        "end_time": str(end_timestamp),
        "bucket_width": "1d",
        "group_by": ["model"],
    }
    headers = {
        "Authorization": f"Bearer {OPENAI_ADMIN_KEY}",
        "Content-Type": "application/json",
    }
    if OPENAI_PROJECT_ID:
        headers["OpenAI-Project"] = OPENAI_PROJECT_ID

    try:
        response = requests.get(url, params=params, headers=headers, timeout=30)
        if not response.ok:
            print(f"❌ Error fetching usage data: {response.status_code}")
            print(f"Response: {response.text}")
            sys.exit(1)
        return response.json()
    except Exception as e:
        print(f"❌ Error fetching usage data: {e}")
        sys.exit(1)


def calculate_cost_from_tokens(usage_data: dict) -> tuple[float, int, int]:
    """トークン使用量からコストを計算（gpt-4o-mini料金で統一）

    Returns:
        tuple[float, int, int]: (総コストUSD, 総inputトークン, 総outputトークン)
    """
    total_input_tokens = 0
    total_output_tokens = 0

    for bucket in usage_data.get("data", []):
        for result in bucket.get("results", []):
            total_input_tokens += result.get("input_tokens", 0)
            total_output_tokens += result.get("output_tokens", 0)

    # gpt-4o-mini料金で計算
    input_cost = (total_input_tokens / 1_000_000) * GPT4O_MINI_PRICING["input"]
    output_cost = (total_output_tokens / 1_000_000) * GPT4O_MINI_PRICING["output"]
    total_cost = input_cost + output_cost

    return total_cost, total_input_tokens, total_output_tokens


def format_tokens(tokens: int) -> str:
    """トークン数を読みやすくフォーマット"""
    if tokens >= 1_000_000:
        return f"{tokens / 1_000_000:.2f}M"
    elif tokens >= 1_000:
        return f"{tokens / 1_000:.1f}K"
    return str(tokens)


def send_slack_notification(message: str, is_alert: bool = False):
    """Slack通知を送信"""
    if not SLACK_WEBHOOK_URL:
        print("⚠️  Slack webhook URL not configured")
        return

    color = "#ff0000" if is_alert else "#36a64f"
    title = "🚨 OpenAI APIコストアラート" if is_alert else "📊 OpenAI API使用量レポート"

    payload = {
        "attachments": [
            {
                "color": color,
                "title": title,
                "text": message,
                "footer": "Stock Buddy Monitoring",
                "ts": int(datetime.now().timestamp()),
            }
        ]
    }

    try:
        requests.post(SLACK_WEBHOOK_URL, json=payload, timeout=10)
        print("✅ Slack notification sent")
    except Exception as e:
        print(f"⚠️  Failed to send Slack notification: {e}")


def main():
    if not OPENAI_ADMIN_KEY:
        print("❌ Error: OPENAI_ADMIN_KEY environment variable is not set")
        sys.exit(1)

    today = datetime.now()

    # 月初の場合は先月分をレポート
    if today.day == 1:
        last_month_end = datetime(today.year, today.month, 1) - timedelta(days=1)
        start_of_month = datetime(last_month_end.year, last_month_end.month, 1)
        end_of_day = datetime(
            last_month_end.year, last_month_end.month, last_month_end.day, 23, 59, 59
        )
        period_label = f"{last_month_end.year}-{last_month_end.month:02d} (先月分)"
    else:
        yesterday = today - timedelta(days=1)
        start_of_month = datetime(today.year, today.month, 1)
        end_of_day = datetime(
            yesterday.year, yesterday.month, yesterday.day, 23, 59, 59
        )
        period_label = f"{today.year}-{today.month:02d} (今月)"

    # 予算を円換算
    budget_jpy = MONTHLY_BUDGET_USD * USD_TO_JPY

    print(f"📊 Checking OpenAI API usage: {period_label}")
    print(f"📅 Period: {start_of_month.date()} to {end_of_day.date()}")
    print(f"💰 Monthly budget: ¥{budget_jpy:,.0f} (${MONTHLY_BUDGET_USD})\n")

    # トークン使用量を取得
    usage_data = get_usage_data(
        int(start_of_month.timestamp()), int(end_of_day.timestamp())
    )

    # コストを計算（gpt-4o-mini料金で統一）
    total_cost_usd, input_tokens, output_tokens = calculate_cost_from_tokens(usage_data)
    total_cost_jpy = total_cost_usd * USD_TO_JPY
    usage_percentage = (total_cost_usd / MONTHLY_BUDGET_USD) * 100

    # コンソール出力
    print(f"✅ Estimated cost: ¥{total_cost_jpy:,.0f} (${total_cost_usd:.4f})")
    print(f"📈 Budget usage: {usage_percentage:.1f}%")
    print(f"💵 Remaining: ¥{budget_jpy - total_cost_jpy:,.0f}")
    print(f"📊 Tokens: in={format_tokens(input_tokens)}, out={format_tokens(output_tokens)}")
    print("ℹ️  ※gpt-4o-mini料金で試算\n")

    # Slack通知メッセージ作成
    slack_lines = [
        f"*期間*: {start_of_month.date()} 〜 {end_of_day.date()} ({period_label})",
        f"*推定コスト*: ¥{total_cost_jpy:,.0f}",
        f"*予算*: ¥{budget_jpy:,.0f}",
        f"*使用率*: {usage_percentage:.1f}%",
        f"*トークン*: in={format_tokens(input_tokens)}, out={format_tokens(output_tokens)}",
        "",
        f"_※gpt-4o-mini料金で試算（$1=¥{USD_TO_JPY}）_",
    ]

    slack_message = "\n".join(slack_lines)

    # アラート判定
    if usage_percentage >= 100:
        print("🚨 CRITICAL: Budget exceeded!")
        send_slack_notification(slack_message + "\n\n⚠️ *予算を超過しました！*", True)
        sys.exit(1)
    elif usage_percentage >= 80:
        print("⚠️  WARNING: 80% of budget used")
        send_slack_notification(slack_message + "\n\n⚠️ 予算の80%に達しました", True)
    else:
        print("✅ Usage is within normal range")
        send_slack_notification(slack_message, False)


if __name__ == "__main__":
    main()

File: scripts/github-actions/check-openai-usage.ts
#!/usr/bin/env npx tsx
/**
 * OpenAI API使用量チェックスクリプト
 *
 * GitHub Actionsから毎日実行され、OpenAI APIの実際のコストを取得します。
 * 予算の80%を超えた場合はアラートを出力します。
 *
 * Costs APIを使用して、OpenAIが計算した実際の請求額を取得します。
 */

// 環境変数
const OPENAI_ADMIN_KEY = process.env.OPENAI_ADMIN_KEY
const OPENAI_PROJECT_ID = process.env.OPENAI_PROJECT_ID
const SLACK_WEBHOOK_URL = process.env.OPENAI_SLACK_WEBHOOK_URL
const MONTHLY_BUDGET_USD = parseFloat(process.env.MONTHLY_BUDGET_USD || "50")

interface CostsResult {
  amount?: {
    value?: string | number
  }
}

interface CostsBucket {
  results?: CostsResult[]
  amount?: {
    value?: string | number
  }
}

interface CostsResponse {
  data?: CostsBucket[]
}

async function getCostsData(startTimestamp: number, endTimestamp: number): Promise<CostsResponse> {
  const url = new URL("https://api.openai.com/v1/organization/costs")
  url.searchParams.set("start_time", startTimestamp.toString())
  url.searchParams.set("end_time", endTimestamp.toString())
  url.searchParams.set("bucket_width", "1d")

  try {
    const response = await fetch(url.toString(), {
      method: "GET",
      headers: {
        Authorization: `Bearer ${OPENAI_ADMIN_KEY}`,
        "Content-Type": "application/json",
        "OpenAI-Project": OPENAI_PROJECT_ID!,
      },
      signal: AbortSignal.timeout(30000),
    })

    if (!response.ok) {
      const text = await response.text()
      console.error(`❌ Error fetching costs data: ${response.status}`)
      console.error(`Response: ${text}`)
      process.exit(1)
    }

    return await response.json()
  } catch (error) {
    console.error(`❌ Error fetching costs data: ${error}`)
    process.exit(1)
  }
}

function calculateTotalCost(costsData: CostsResponse): number {
  let totalCost = 0.0

  if (!costsData.data) {
    return totalCost
  }

  for (const bucket of costsData.data) {
    if (bucket.results) {
      for (const result of bucket.results) {
        const amountValue = result.amount?.value ?? 0
        const costUsd = typeof amountValue === "string" ? parseFloat(amountValue) : amountValue
        totalCost += costUsd || 0
      }
    } else if (bucket.amount) {
      const amountValue = bucket.amount.value ?? 0
      const costUsd = typeof amountValue === "string" ? parseFloat(amountValue) : amountValue
      totalCost += costUsd || 0
    }
  }

  return totalCost
}

async function sendSlackNotification(message: string, isAlert: boolean = false): Promise<void> {
  if (!SLACK_WEBHOOK_URL) {
    console.log("⚠️  Slack webhook URL not configured, skipping notification")
    return
  }

  const color = isAlert ? "#ff0000" : "#36a64f"
  const payload = {
    attachments: [
      {
        color,
        title: isAlert ? "🚨 OpenAI APIコストアラート" : "📊 OpenAI API使用量レポート",
        text: message,
        footer: "Stock Buddy Monitoring",
        ts: Math.floor(Date.now() / 1000),
      },
    ],
  }

  try {
    const response = await fetch(SLACK_WEBHOOK_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(10000),
    })

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`)
    }
    console.log("✅ Slack notification sent")
  } catch (error) {
    console.log(`⚠️  Failed to send Slack notification: ${error}`)
  }
}

function formatDate(date: Date): string {
  return date.toISOString().split("T")[0]
}

async function main(): Promise<void> {
  if (!OPENAI_ADMIN_KEY) {
    console.error("❌ Error: OPENAI_ADMIN_KEY environment variable is not set")
    process.exit(1)
  }

  if (!OPENAI_PROJECT_ID) {
    console.error("❌ Error: OPENAI_PROJECT_ID environment variable is not set")
    process.exit(1)
  }

  const today = new Date()
  let startOfMonth: Date
  let endOfDay: Date
  let periodLabel: string

  if (today.getDate() === 1) {
    // 月初の場合は先月のデータを取得
    const lastMonth = new Date(today)
    lastMonth.setDate(0) // 先月の最終日
    startOfMonth = new Date(lastMonth.getFullYear(), lastMonth.getMonth(), 1, 0, 0, 0, 0)
    endOfDay = new Date(lastMonth.getFullYear(), lastMonth.getMonth(), lastMonth.getDate(), 23, 59, 59, 999)
    periodLabel = `${lastMonth.getFullYear()}-${String(lastMonth.getMonth() + 1).padStart(2, "0")} (先月分)`
  } else {
    // 月初以外は今月のデータを昨日まで取得
    const yesterday = new Date(today)
    yesterday.setDate(today.getDate() - 1)
    startOfMonth = new Date(today.getFullYear(), today.getMonth(), 1, 0, 0, 0, 0)
    endOfDay = new Date(yesterday.getFullYear(), yesterday.getMonth(), yesterday.getDate(), 23, 59, 59, 999)
    periodLabel = `${today.getFullYear()}-${String(today.getMonth() + 1).padStart(2, "0")} (今月)`
  }

  const startTimestamp = Math.floor(startOfMonth.getTime() / 1000)
  const endTimestamp = Math.floor(endOfDay.getTime() / 1000)

  console.log(`📊 Checking OpenAI API costs: ${periodLabel}`)
  console.log(`📅 Period: ${formatDate(startOfMonth)} to ${formatDate(endOfDay)}`)
  console.log(`🎯 Project: Stock Buddy (${OPENAI_PROJECT_ID})`)
  console.log(`💰 Monthly budget: $${MONTHLY_BUDGET_USD}`)
  console.log()

  // コストデータを取得
  const costsData = await getCostsData(startTimestamp, endTimestamp)

  // 実際のコストを集計
  const totalCost = calculateTotalCost(costsData)
  const usagePercentage = (totalCost / MONTHLY_BUDGET_USD) * 100

  // 結果を表示
  console.log(`✅ Total cost this month: $${totalCost.toFixed(2)}`)
  console.log(`📈 Budget usage: ${usagePercentage.toFixed(1)}%`)
  console.log(`💵 Remaining budget: $${(MONTHLY_BUDGET_USD - totalCost).toFixed(2)}`)
  console.log()

  // Slack通知メッセージを作成
  let slackMessage = `
*期間*: ${formatDate(startOfMonth)} 〜 ${formatDate(endOfDay)} (${periodLabel})
*プロジェクト*: Stock Buddy
*総コスト*: $${totalCost.toFixed(2)}
*予算*: $${MONTHLY_BUDGET_USD}
*使用率*: ${usagePercentage.toFixed(1)}%
*残り*: $${(MONTHLY_BUDGET_USD - totalCost).toFixed(2)}
`

  // アラート判定とSlack通知
  if (usagePercentage >= 100) {
    console.log("🚨 CRITICAL: Budget exceeded! Immediate action required.")
    console.log("   - Consider reducing AI analysis frequency")
    console.log("   - Review API usage patterns")
    slackMessage += "\n⚠️ *予算を超過しました！早急な対応が必要です*"
    await sendSlackNotification(slackMessage, true)
    process.exit(1)
  } else if (usagePercentage >= 80) {
    console.log("⚠️  WARNING: 80% of monthly budget used")
    console.log("   - Monitor usage closely")
    console.log("   - Consider optimizing prompts")
    slackMessage += "\n⚠️ 予算の80%に達しました。使用量を注視してください"
    await sendSlackNotification(slackMessage, true)
  } else if (usagePercentage >= 50) {
    console.log("ℹ️  INFO: 50% of monthly budget used")
    slackMessage += "\nℹ️ 予算の50%に達しました"
    await sendSlackNotification(slackMessage, false)
  } else {
    console.log("✅ Usage is within normal range")
    await sendSlackNotification(slackMessage, false)
  }

  // 詳細データをMarkdown形式で出力（GitHub Actions Summaryで使用）
  console.log()
  console.log("## Usage Details")
  console.log(`- **Period**: ${formatDate(startOfMonth)} to ${formatDate(endOfDay)} (${periodLabel})`)
  console.log(`- **Project**: Stock Buddy (${OPENAI_PROJECT_ID})`)
  console.log(`- **Total Cost**: $${totalCost.toFixed(2)}`)
  console.log(`- **Budget**: $${MONTHLY_BUDGET_USD}`)
  console.log(`- **Usage**: ${usagePercentage.toFixed(1)}%`)
  console.log(`- **Remaining**: $${(MONTHLY_BUDGET_USD - totalCost).toFixed(2)}`)
}

main().catch((error) => {
  console.error(error)
  process.exit(1)
})

export {}
